File: backend/src/controllers/transcriptController.py
#!/usr/bin/env python3
import sys
import json
import traceback

try:
    from youtube_transcript_api import YouTubeTranscriptApi
    from youtube_transcript_api.formatters import TextFormatter
    # Additional imports if needed
except ImportError:
    print(json.dumps({
        "success": False,
        "error": "Required libraries not installed. Please install with: pip install youtube-transcript-api"
    }))
    sys.exit(1)

# Import your models and DB connection
from models.postModel import Post
from middleware.authMiddleware import protect

# Helper function to extract video ID from YouTube URL
def extract_video_id(url):
    """Extract the video ID from a YouTube URL"""
    pattern = r'(?:v=|\/)([0-9A-Za-z_-]{11}).*'
    match = re.search(pattern, url)
    return match.group(1) if match else None

def get_transcript(video_id):
    """Get transcript for a YouTube video."""
    try:
        # Get available transcripts
        transcript_list = YouTubeTranscriptApi.list_transcripts(video_id)
        
        # Try to get English transcript first
        try:
            transcript = transcript_list.find_transcript(['en'])
        except:
            # If English not available, get the first available transcript
            transcript = transcript_list.find_generated_transcript(['en'])
            if not transcript:
                # Get any transcript (auto-generated if needed)
                transcript = list(transcript_list)[0]
        
        # Get transcript data
        transcript_data = transcript.fetch()
        
        # Format transcript as plain text
        formatter = TextFormatter()
        plain_text = formatter.format_transcript(transcript_data)
        
        # Get transcript info
        language = transcript.language
        language_code = transcript.language_code
        is_generated = transcript.is_generated
        
        return {
            "success": True,
            "transcript": plain_text,
            "language": language,
            "language_code": language_code,
            "is_generated": is_generated,
            "videoId": video_id
        }
        
    except Exception as e:
        error_details = traceback.format_exc()
        return {
            "success": False,
            "error": str(e),
            "details": error_details,
            "videoId": video_id
        }

def main():
    """Main function to handle command line arguments."""
    if len(sys.argv) < 2:
        print(json.dumps({
            "success": False,
            "error": "Video ID is required"
        }))
        sys.exit(1)
    
    video_id = sys.argv[1]
    result = get_transcript(video_id)
    print(json.dumps(result))

if __name__ == "__main__":
    main()

# Save video with transcript
def save_video_transcript():
    """Save a video transcript to the database"""
    try:
        data = request.get_json()
        video_url = data.get('videoUrl')
        video_id = data.get('videoId')
        title = data.get('title', f'YouTube Video: {video_id}')
        transcript = data.get('transcript')
        
        # Get user ID from auth middleware
        user_id = request.user.get('_id') if request.user else None
        
        if not all([video_url, video_id, transcript]):
            return jsonify({
                'success': False,
                'message': 'Video URL, ID, and transcript are required'
            }), 400
        
        # Check if we already have this video
        existing_video = Post.find_one({
            'platformPostId': video_id, 
            'platform': 'youtube',
            'userId': user_id
        })
        
        if existing_video:
            # Update existing video
            updated_post = Post.update_one(
                {'_id': existing_video['_id']},
                {
                    '$set': {
                        'content': transcript,
                        'title': title,
                        'updatedAt': datetime.now()
                    }
                }
            )
            
            return jsonify({
                'success': True,
                'message': 'Video transcript updated successfully',
                'post': {
                    '_id': str(existing_video['_id']),
                    'title': title,
                    'content': transcript,
                    'platformPostId': video_id
                }
            }), 200
        else:
            # Create new video entry
            new_post = {
                'title': title,
                'content': transcript,
                'status': 'draft',
                'platform': 'youtube',
                'platformPostId': video_id,
                'videoUrl': video_url,
                'userId': user_id,
                'createdAt': datetime.now(),
                'updatedAt': datetime.now()
            }
            
            result = Post.insert_one(new_post)
            new_post['_id'] = str(result.inserted_id)
            
            return jsonify({
                'success': True,
                'message': 'Video transcript saved successfully',
                'post': new_post
            }), 201
            
    except Exception as e:
        print(f"Error saving video transcript: {str(e)}")
        return jsonify({
            'success': False,
            'message': f'Failed to save video transcript: {str(e)}'
        }), 500 

File: backend/src/controllers/transcriptController.js
const { spawn } = require('child_process');
const path = require('path');
const asyncHandler = require('express-async-handler');
const axios = require('axios');

// Cache for storing transcripts to avoid repeated YouTube API calls
const transcriptCache = new Map();

// Helper to add delay to avoid rate limiting
const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

/**
 * @desc    Get transcript for a YouTube video
 * @route   POST /api/transcript
 * @access  Public
 */
const getTranscript = asyncHandler(async (req, res) => {
  const { videoId } = req.body;

  if (!videoId) {
    res.status(400);
    throw new Error('Video ID is required');
  }

  console.log(`Attempting to get transcript for video ID: ${videoId}`);

  // Check cache first
  if (transcriptCache.has(videoId)) {
    console.log(`Using cached transcript for video ID: ${videoId}`);
    return res.status(200).json(transcriptCache.get(videoId));
  }

  try {
    const pythonScript = path.join(__dirname, '..', 'transcript_fetcher.py');
    
    // Use platform-specific Python command
    const isWindows = process.platform === "win32";
    const pythonCommand = isWindows
      ? 'C:\\Users\\hp\\AppData\\Local\\Programs\\Python\\Python313\\python.exe' // Local Windows
      : 'python3'; // Render/Linux
    
    console.log(`Using Python command: ${pythonCommand} for platform: ${process.platform}`);
    console.log(`Python script path: ${pythonScript}`);
    
    // Execute the Python script and pass the videoId as an argument
    const pythonProcess = spawn(pythonCommand, [pythonScript, videoId]);
    
    let transcriptData = '';
    let errorData = '';

    // Collect data from stdout
    pythonProcess.stdout.on('data', (data) => {
      transcriptData += data.toString();
    });

    // Collect any error output
    pythonProcess.stderr.on('data', (data) => {
      errorData += data.toString();
      console.error(`Python stderr: ${data}`);
    });

    // When the process completes
    pythonProcess.on('close', (code) => {
      if (code !== 0) {
        console.error(`Python process exited with code ${code}`);
        console.error(`Error output: ${errorData}`);
        
        // Check for rate limiting errors
        if (errorData.includes('429') || errorData.includes('Too Many Requests')) {
          return res.status(429).json({ 
            success: false, 
            error: 'YouTube API rate limit exceeded. Please try again later.',
            details: 'Too many requests to YouTube. This is a temporary issue.'
          });
        }
        
        return res.status(500).json({ 
          success: false, 
          error: 'Failed to get transcript', 
          details: errorData 
        });
      }

      try {
        const result = JSON.parse(transcriptData);
        
        if (!result.success) {
          console.log(`Transcript not found for video ID ${videoId}: ${result.error}`);
          return res.status(404).json(result);
        }
        
        console.log(`Successfully retrieved transcript for video ID ${videoId}, length: ${result.transcript.length} chars`);
        
        // Cache the successful result
        transcriptCache.set(videoId, result);
        
        return res.status(200).json(result);
      } catch (error) {
        console.error('Error parsing transcript data:', error);
        console.error('Raw data:', transcriptData);
        return res.status(500).json({ 
          success: false, 
          error: 'Failed to parse transcript data',
          details: error.message
        });
      }
    });
    
    // Handle process errors (e.g. if Python executable not found)
    pythonProcess.on('error', (err) => {
      console.error('Failed to start Python process:', err);
      return res.status(500).json({
        success: false,
        error: 'Failed to start Python process',
        details: err.message
      });
    });
  } catch (error) {
    console.error('Error executing Python script:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Server error while getting transcript',
      details: error.message
    });
  }
});

/**
 * @desc    Get transcript directly using YouTube API
 * @route   POST /api/transcript/direct
 * @access  Public
 * @note    This is a fallback for when the Python method fails
 */
const getTranscriptDirect = asyncHandler(async (req, res) => {
  const { videoId } = req.body;
  
  if (!videoId) {
    res.status(400);
    throw new Error('Video ID is required');
  }
  
  console.log(`Attempting to get direct transcript for video ID: ${videoId}`);
  
  // Check cache first
  if (transcriptCache.has(videoId)) {
    console.log(`Using cached transcript for video ID: ${videoId}`);
    return res.status(200).json(transcriptCache.get(videoId));
  }
  
  try {
    // Add small delay to avoid rate limiting
    await delay(500);
    
    // Try to fetch the caption track
    const captionUrl = `https://www.youtube.com/api/timedtext?lang=en&v=${videoId}`;
    const response = await axios.get(captionUrl);
    let transcript = response.data;
    
    if (!transcript || transcript.trim() === '') {
      console.log(`No direct transcript available for video ID ${videoId}`);
      return res.status(404).json({ 
        success: false, 
        error: 'No transcript available for this video' 
      });
    }
    
    // Simple XML parsing to extract text
    transcript = transcript
      .replace(/<text[^>]*>/g, '')
      .replace(/<\/text>/g, ' ')
      .replace(/&amp;/g, '&')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&quot;/g, '"')
      .replace(/&#39;/g, "'")
      .replace(/\n/g, ' ');
    
    console.log(`Successfully retrieved direct transcript for video ID ${videoId}, length: ${transcript.length} chars`);
    
    const result = {
      success: true,
      transcript: transcript,
      language: 'English',
      language_code: 'en',
      is_generated: true
    };
    
    // Cache the successful result
    transcriptCache.set(videoId, result);
    
    return res.status(200).json(result);
  } catch (error) {
    console.error('Error fetching direct transcript:', error);
    
    // Check for rate limiting
    if (error.response && error.response.status === 429) {
      return res.status(429).json({
        success: false,
        error: 'YouTube API rate limit exceeded. Please try again later.',
        details: 'Too many requests to YouTube. This is a temporary issue.'
      });
    }
    
    return res.status(500).json({ 
      success: false, 
      error: 'Failed to fetch direct transcript', 
      details: error.message 
    });
  }
});

module.exports = {
  getTranscript,
  getTranscriptDirect
}; 